proposals/0530-strict-box-denesting.rst

```diff
@@ -0,0 +1,527 @@
+
+The main effect of this proposal would be to cause some Haskell
+programs to run faster and use less memory.
+
+This proposal interacts nontrivially with the ``dataToTag#`` primop,
+which currently always evaluates its argument and returns the index of
+the data constructor (if it exists) used to produce the resulting
+evaluated heap object.  Since ``-fstrict-box-denesting`` affects the
+set of things that produce new evaluated heap objects, it can affect
+the result of a call to the current implementation of
+``dataToTag#``.  For example, ``dataToTag# (MkExample1 [True])`` with
+its current implementation may return either
+
+* the index ``0#`` of the ``MkExample1`` constructor, or
+* the index ``1#`` of the ``(:)`` constructor of the underlying
+  heap object ``(:) True []``
+
+\...depending on whether or not the ``-fstrict-box-denesting`` flag is
+active when in the module containing ``Example1`` is compiled.  This
+proposal suggests to retain the current behavior of ``dataToTag#``
+regardless of the ``-fstrict-box-denesting`` flag by allowing a
+different implementation to be used for types with a denestable
+constructor, via run-time evidence of a ``DataToTag`` typeclass
+constraint.
+
+
+
+Costs and Drawbacks
+-------------------
+1. Today, most or all of GHC's boxed primitive types such as
+   ``ByteArray#`` cannot be safely stored in a value with lifted type,
+   as they are zero-tagged pointers (so they will be entered when
+   evaluated) but entering them results in a panic.  These panics have
+   been useful for GHC developers in the past, but are incompatible
+   with this proposal as it is currently written, since it allows
+   a wrapper such as ``ShortByteString`` to be represented at runtime
+   by its underlying ``ByteArray#``.
+
+   Since entering these objects involves an indirection that would
+   largely defeat the purpose of denesting their lifted wrappers, this
+   must be dealt with by changing the proper tag for their pointers.
+   The panicking entry code can very likely be kept, but will now
+   detect a different class of bugs: those where pointer tags are not
+   preserved or where tag inference incorrectly concludes that a
+   pointer must be untagged.  (See also `GHC issue 21792
+   <https://gitlab.haskell.org/ghc/ghc/-/issues/21792>`_\ .)
+
+2. This proposal calls for a breaking change to the type of the
+   ``dataToTag#`` primitive, making it not applicable in some
+   situations where it can currently be used.
+
+3. The heap objects elided at runtime by ``-fstrict-box-denesting``
+   may occasionally provide useful information for debugging or profiling.
+
+
+The implementation of this proposal is not expected to pose major
+challenges or incur much ongoing maintenance cost.  It consists of
+three largely independent parts:
+
+* Wire in and give special solving behavior to a ``DataToTag`` class.
+  This is expected to be very similar to the recent work that
+  implemented the ``WithDict`` class.
```

Comment:
The solving is about 20 lines in my WIP implementation at [ghc!8912](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8912).

Reply:
Ah, GHC issue 21575 is not the one I thought you meant. It exists because `withDict` doesn't play nice with the specializer and is still open because the implementation of `withDict` in ghc-9.4.2 still has this problem. (The implementation in ghc-HEAD is not known to have this problem.) It's unrelated to the recently added solving behavior for `WithDict`, which is working as intended (to my knowledge).

Reply:
Yes, you're probably right; it's not the constraint resolution that makes `WithDict` buggy, it's the loss of coherence (e.g., `Note [Type determines value]`).

Plus, Simon does not seem to worry either.